02-wrangling.Rmd

```diff
@@ -256,55 +320,182 @@ lang_no_delimiter <- separate(lang_messy_longer,
+
+We can see the populations of the five cities by filtering the data
+using the `%in%` operator. The `%in%` operator is used to see if an element 
+belongs to a vector. Here we are filtering if any of the five cities 
+("Toronto", "Montréal", "Vancouver", "Calgary", "Edmonton") match the rows in the `region`
+column.
+```{r}
+filter(region_data, region %in% c("Toronto", "Montréal", "Vancouver", "Calgary", "Edmonton")) 
+```
+
+> **What's the difference between `==` and `%in%`?**
+> Let's suppose we have two vectors, `vector1` and `vector2`. If you type 
+> `vector1 == vector2` into R it
+will compare the vectors element by element. R checks if the first element of 
+`vector1` equals the first element of `vector2`, the second element of `vector1`
+equals the second element of `vector2` etc. 
+> On the other hand, `vector1 %in% vector2`
+compares the first element of `vector1` to all the elements in `vector2`. 
+Then the second element of `vector1` is then compared to all the elements in `vector2` etc. Notice the difference between `==` and `%in%` in the example below. 
+ 
+```{r}
+c("Vancouver", "Toronto") == c("Toronto", "Vancouver")
+c("Vancouver", "Toronto") %in% c("Toronto", "Vancouver")
+```
+
+```{r, include = F}
+city_popn <- region_data |> filter(region %in% c("Toronto", "Montréal", "Vancouver", "Calgary", "Edmonton")) |> arrange(desc(population) ) |> pull(population)
+rep_popn <- rep(city_popn, 2)
+```
+
+We can create a vector of the city populations. 
+Then we use `mutate` to calculate
+the proportion of people reporting a given language as their primary language at home for all the languages in the `official_langs` data set by dividing the column by our vector of city populations. We give the mutated column a new name `most_at_home_proportion`, which will create a new column in our data set. 
+```{r}
+city_populations <- c(5928040, 4098927, 2463431, 1392609, 1321426)
+
+official_langs <- mutate(official_langs,
+                           most_at_home_proportion = most_at_home / city_populations)
+
+official_langs
+```
+
+> Notice that we divided a vector with 10 elements (`most_at_home`) by a vector with 
+only 5 elements (`city_populations`). What gives? Arithmetic operations of vectors are performed *memberwise* meaning the first element of `most_at_home` will be divided by the first element 
+of `city_populations`, the second element of `most_at_home` will be divided by the second element of `city_populations` and so on. When two vectors have different lengths, the 
+shorter one will be recycled to match the longer one. Here the elements of
+`city_populations` will be repeated once to match the length of `most_at_home`.
```

Comment:
I am quite concerned about teaching vector recycling... I think it might be good to teach that R does this, R doing this makes sense when vectors are of length 1, but can easily become an error-prone feature, when vectors are longer than length 1. 

In my own data wrangling, I would probably do some sort of `case_when` or `join`, but that too advanced for here... 

Actually, `case_when` might be good - as it goes inside `mutate`... but if we do this, we should teach a more vanilla `mutate` first...

Reply:
Yes, that is a very good point! I was trying to pick the "easiest" thing to do that wouldn't involve explaining a lot of new functions, but you're right that we probably don't want to show students this and then they will do it and its not best practice. So I left it in so we have a more vanilla mutate example, but added "this is actually a bad way to do it" . Then I show the better method with `case_when` and then at the end say there are multiple ways to accomplish the same thing!